[AC-1479][BEEEP] Refactor ConfigService to improve observable usage

eliykat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
Fix a race condition in `ConfigService`, where `getFeatureFlagBool` (and other similar methods) would always return `false` when called immediately on app start, because:
* `stateService` did not have any existing stored value
* `serverConfig$` had not emitted a value yet because the api call had not completed

More generally, I thought this service could be improved - `serverConfig$` seems like it should be the source of truth here, but it's not because it's not seeded with the initial value from `stateService`, which can only be accessed via the `getFeatureFlag` methods. I also wanted a feature flag observable, rather than wrapping an observable in a promise interface.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

**ConfigService:**
- the main work is now done in the constructor: there's an observable for the `stateService` call, an observable for the http call, and they both emit their values via the exposed `serverConfig$` observable
- there's a method to give you a feature flag observable, `getFeatureFlag$`, with a promise-based wrapper for backwards compatibility
- removed the `featureFlagBool` etc methods which were just wrappers around a generic method. I'm not sure why these were required, generics are not very cumbersome and (for example) it complicated the route guard unnecessarily

**Tests**
The tests aren't the most elegant because I couldn't figure out TestScheduler.

**Other changes**
Other changes are minor and I've intentionally tried to change as little as possible.

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

eliykat commented:
I'll fix linting and typechecking errors tomorrow.

Hinton commented:
We already have some work planned for the config service in https://bitwarden.atlassian.net/browse/PM-2304.

eliykat commented:
Good point, I'll make sure to check with the team first next time. This was mostly BEEEP time because I thought it would be a nice refactor, but if it conflicts with or gets in the way of other planned work I don't mind closing it.

MGibson1 commented on libs/common/src/platform/abstractions/config/config.service.abstraction.ts:
```diff
@@ -6,8 +6,7 @@ import { ServerConfig } from "./server-config";
-  getFeatureFlagNumber: (key: FeatureFlag, defaultValue?: number) => Promise<number>;
+  getFeatureFlag$: <T>(key: FeatureFlag, defaultValue?: T) => Observable<T>;
+  getFeatureFlag: <T>(key: FeatureFlag, defaultValue?: T) => Promise<T>;
+  fetchServerConfig: () => void;
```
This is a weird interface to me. I would expect a `fetch` method to be async and awaitable, but this isn't. I'm not sure this is what we want, yet, but I would at least rename this to `triggerServerConfigFetch`.

eliykat replied:
It's not really what we want, it's there for backwards compatibility only. We should be able to subscribe to an observable to kick off a new fetch from server - like we do with `environmentService.urls`. iirc, it's only used for sync at the moment, and we don't have a "sync completed" observable we can subscribe to.

I'll rename it as you suggest though.

MGibson1 replied:
I'm confused, I don't see any behavior like that in environment service. There's an observable to grab from storage again on account change.

We actually do have plans for a sync completed observable -- see `SyncNotifierService`. It's not hooked up right now because there are concurrency issues with state due to deep objects

eliykat replied:
I mean in ConfigService, we subscribe to `environmentService.urls` to trigger a new config fetch: https://github.com/bitwarden/clients/blob/config-service-refactor/libs/common/src/platform/services/config/config.service.ts#L55

I did spot `SyncNotifierService`, and ideally we'd subscribe to `syncNotifierService.sync$` in that same `merge`, but as you say it's not used yet.

MGibson1 replied:
ah -- I see. That makes more sense now, thank you.

MGibson1 commented on libs/common/src/platform/abstractions/config/config.service.abstraction.ts:
```diff
@@ -6,8 +6,7 @@ import { ServerConfig } from "./server-config";
-  getFeatureFlagBool: (key: FeatureFlag, defaultValue?: boolean) => Promise<boolean>;
-  getFeatureFlagString: (key: FeatureFlag, defaultValue?: string) => Promise<string>;
-  getFeatureFlagNumber: (key: FeatureFlag, defaultValue?: number) => Promise<number>;
+  getFeatureFlag$: <T>(key: FeatureFlag, defaultValue?: T) => Observable<T>;
```
_technically_ this not having a default value of `""` is a breaking change. The others will coerce properly, but this won't.

I don't think it's used anywhere and I think this choice is better, so as long as you concur, this seems good. Just wanted to make sure we thought about it.

eliykat replied:
I hadn't thought about it! But you're right, we only use boolean feature flags at the moment, so I agree it's not an issue.

MGibson1 commented on libs/common/src/platform/abstractions/config/config.service.abstraction.ts:
```diff
@@ -6,8 +6,7 @@ import { ServerConfig } from "./server-config";
-  getFeatureFlagBool: (key: FeatureFlag, defaultValue?: boolean) => Promise<boolean>;
-  getFeatureFlagString: (key: FeatureFlag, defaultValue?: string) => Promise<string>;
-  getFeatureFlagNumber: (key: FeatureFlag, defaultValue?: number) => Promise<number>;
+  getFeatureFlag$: <T>(key: FeatureFlag, defaultValue?: T) => Observable<T>;
+  getFeatureFlag: <T>(key: FeatureFlag, defaultValue?: T) => Promise<T>;
```
I'd like to see some type restrictions on `T`
```suggestion
  getFeatureFlag$: <T extends boolean | number | string>(key: FeatureFlag, defaultValue?: T) => Observable<T>;
  getFeatureFlag: <T extends boolean | number | string>(key: FeatureFlag, defaultValue?: T) => Promise<T>;
```

eliykat replied:
Great suggestion! done

MGibson1 commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -13,76 +24,67 @@ import { ServerConfigData } from "../../models/data/server-config.data";
 export class ConfigService implements ConfigServiceAbstraction {
   protected _serverConfig = new BehaviorSubject<ServerConfig | null>(null);
```
Should not initialize this and just initialize as the result from your constructor's call to `stateService.getServerConfig`. That way the value is always the best info we have.

eliykat replied:
I do wonder whether we rely on BehaviorSubjects too much, I see a lot of `null` initializations around here.

We can't initialize with an async value, but we can use a `ReplaySubject` instead so that it just doesn't emit any value until the first real value is coming down the pike.

This does simplify the tests - we can drop all of the `// skip initial null value` stuff. However, it means I can't use `this._serverConfig.getValue()` to check for a current value and avoid overwriting it from storage (basically avoiding a race condition between storage & server). 

I refactored it to `concat` the two observables together in the expected order - so we first hit storage, then the server, but now `configFromServer$` is never emitting, and I can't figure out why. `configFromStorage$` should always complete (which it seems to), but then `configFromServer$` never fires. Any suggestions here would be welcome because I couldn't figure it out. (or if there's a different approach altogether)

eliykat replied:
Actually, I'm not sure this is a good idea. Assume there's no stored serverConfig, so that null value is filtered out and not emitted. Also assume the client is offline. Then no value is emitted and lastValueFrom will never resolve.

MGibson1 replied:
That just means there needs to be a final fallback that emits defaults should storage and server fail, right?

eliykat replied:
This should be resolved now. Let me know what you think of the latest changes.

MGibson1 commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -13,76 +24,67 @@ import { ServerConfigData } from "../../models/data/server-config.data";
-      .subscribe((serverConfig) => {
-        this._serverConfig.next(serverConfig);
-      });
+    // Get config from storage on initial load
+    from(this.stateService.getServerConfig())
+      .pipe(
+        filter((data) => data != null && this._serverConfig.getValue() == null),
+        map((data) => new ServerConfig(data))
+      )
+      .subscribe((config) => this._serverConfig.next(config));
```
Should init `_serverConfig` with this rather than `null`

eliykat replied:
The latest changes should address this.

MGibson1 reviewed (changes requested):
This is a definite improvement.

I did have a few things that I think will improve further that should be merged as a part of this.

eliykat commented:
I've reverted `ReplaySubject` changes because they risked breaking the general approach here. Happy to revisit if you have any specific recommendations, otherwise initializing with `null` still seems useful to me to make sure we always have some value.

I've also merged in Oscar's changes from #5520 and the latest changes from `master`, including the EU datacenter changes (`setCloudWebVaultUrl`).

eliykat commented:
I think the only outstanding feedback item was this: https://github.com/bitwarden/clients/pull/5602#discussion_r1246564268

That has been addressed by https://github.com/bitwarden/clients/pull/5602/commits/c687e3ea6f7139381ac89f7c6aaf4da281e59eaf.

eliykat commented:
Note to self: the feature flags page in contributing-docs will need updating after this is merged as it refers to the specific interfaces.

MGibson1 commented on apps/web/src/app/vault/individual-vault/vault.component.ts:
```diff
@@ -879,17 +879,9 @@ export class VaultComponent implements OnInit, OnDestroy {
   async isLowKdfIteration() {
-    const showLowKdfEnabled = await this.configService.getFeatureFlag<boolean>(
-      FeatureFlag.DisplayLowKdfIterationWarningFlag
-    );
```
so this flag was removed?

eliykat replied:
Yes; if you look at the overall diff for the PR it does not show as a change.

MGibson1 commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -13,36 +14,47 @@ import {
+
 export class ConfigService implements ConfigServiceAbstraction {
-  protected _serverConfig = new BehaviorSubject<ServerConfig | null>(null);
+  protected _serverConfig = new ReplaySubject<ServerConfig | null>(1);
```
Isn't this value never `null` now?

eliykat replied:
This can emit `null` if no value has been saved:

```ts
    from(this.stateService.getServerConfig())
      .pipe(map((data) => (data == null ? null : new ServerConfig(data))))
      .subscribe((config) => this._serverConfig.next(config));
```

and we do still want it to emit _something_ to trigger other values downstream (e.g. so that `getFeatureFlag$` can map the default value).

MGibson1 reviewed (commented):
A few comments, but I think this is looking done

eliykat commented:
I just resolved conflicts. Note that none of the conflicted changes to `ConfigService` needed to be preserved because they all addressed issues that are already fixed in this PR. For comparison, those changes were made in the TDE PR here: https://github.com/bitwarden/clients/commit/5665576147a4fbf30024f56358d424ff26a1a131#diff-ccdda88927978910e31f0bb578effd4abcd62d5c7f7971146b74748a58f55883

MGibson1 commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -1,103 +1,99 @@
-  async getFeatureFlagBool(key: FeatureFlag, defaultValue = false): Promise<boolean> {
-    return await this.getFeatureFlag(key, defaultValue);
+    // Get config from storage on initial load
+    from(this.stateService.getServerConfig())
```
I'm sorry this is coming from out of the blue, but lately I've been fighting with calls to promises on `StateService` in the constructor of services. It doesn't allow state service to be initialized prior to attempting to call methods on it.

The better pattern (for now) is to have an `init()` method that is called in `init.service` for the various clients _after_ state service has been initialized

eliykat replied:
No worries, just the normal `stateService` fun then 😀

MGibson1 commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -1,103 +1,99 @@
-    return await this.getFeatureFlag(key, defaultValue);
+    // Get config from storage on initial load
+    from(this.stateService.getServerConfig())
+      .pipe(map((data) => (data == null ? null : new ServerConfig(data))))
```
I'd still prefer to emit a default rather than null here. The expected result should be a warning presented that Bitwarden is down or something should we not get a config back, not just try and continue forward without config information.

I'm willing to move this point to a backlog ticket, though.

eliykat replied:
Sounds like there are 2 parts to that:
1. create a default `ServerConfig` object to act as the default value so we can guarantee the config itself is never null (I hate nulls, so sure)
2. surface an error toast if the server cannot be reached

Both are reasonable but I want to call them out of scope here. I'm not sure what default values should be for (1) and (2) requires design feedback.

That work would be owned by Platform, do you mind creating the ticket for it?

MGibson1 reviewed (changes requested):
Sorry about the new item -- I do believe this is the last, though 🕺

MGibson1 commented on libs/common/src/platform/abstractions/config/config.service.abstraction.ts:
```diff
@@ -6,6 +6,7 @@ import { Region } from "../environment.service";
 import { ServerConfig } from "./server-config";
 
 export abstract class ConfigServiceAbstraction {
+  init: () => void;
```
I would prefer leaving init off of the public interface and requiring a concrete type in the init services.

This isn't something that should be public knowledge to most consumers of this service

cturnbull-bitwarden commented on apps/browser/src/popup/services/init.service.ts:
```diff
@@ -17,7 +18,8 @@ export class InitService {
     private logService: LogServiceAbstraction,
-    private themingService: AbstractThemingService
+    private themingService: AbstractThemingService,
+    private configService: ConfigService
```
:art: This might just be happening on my machine, but when I load the branch locally, I'm met with this console error and a spinning home page:

![image](https://github.com/bitwarden/clients/assets/133619638/704d7211-c9e8-415e-ae1d-56d70d5457ce)

![image](https://github.com/bitwarden/clients/assets/133619638/d7b482ab-c360-4c3a-8d2b-1c278fee1fc3)

This might be because we're instantiating `ConsoleService` directly in each of the `init.service.ts` files, since [jslib-services.module.ts](https://github.com/bitwarden/clients/blob/config-service-refactor/libs/angular/src/services/jslib-services.module.ts#L650) defines the provider to use the abstract class. If that's indeed the issue, then I'd recommend exposing the `init` method by adding it to the abstract class, and then updating the dependencies in each `init.service.ts` to use the abstract class instead.

I tried out those changes in the branch `config-service-refactor-conner` and found that it worked out (is there a way to propose changes in a PR using another branch?). Using that branch, I was able to verify that the billing logic all checks out fine 👍

eliykat replied:
Thank you @cturnbull-bitwarden for actually running the code 😂 

@MGibson1 Conner makes a good point, we can't inject the concrete implementation in the `InitService` constructor without registering that as its own DI token in the services module. But no other class should have a reason to inject the implementation rather than the abstraction, so registering it also seems undesirable.

Adding `init()` to its interface seems like the lesser of two evils, but I don't really mind, what do you want to do?

eliykat replied:
Also @cturnbull-bitwarden 

> is there a way to propose changes in a PR using another branch?

Not really in the same PR, but you can always create a new PR targeting the branch, then link to it from the first PR.

eliykat replied:
@cturnbull-bitwarden I discussed this with @MGibson1 and I've now registered the concrete implementation in the services module, which resolved the problem. I was also missing a call to `init()` in the browser background page. I've tested all 3 clients.

cturnbull-bitwarden replied:
Great! I'll resolve this then

eliykat commented:
Dependency injection and initialization issues have now been fixed per the comment thread above.

In the course of testing, I found that `environmentService.urls` emits when the app is first loaded, which was causing the config endpoint to be hit twice on first load. Also, this behaviour was inconsistent because `environmentService.urls` was a plain old `Subject`, so sometimes it fired before `ConfigService` had a chance to subscribe - but not always.

To fix this, I've changed `environmentService.urls` to be a `ReplaySubject`, so that `ConfigService` will always receive the first emit. We use that to kick off the first fetch from server, instead of when the stored value is emitted. We use `delayWhen` to ensure that the stored value is always emitted first, so that it won't overwrite the value fetched from server.

`environmentService.urls` is only used to kick off `notificationService.init()`, I can't see that changing it to a `ReplaySubject` will cause any regressions there - if anything, it'll avoid the same race condition problem.

jlf0dev reviewed (approved):
Looks good on the Auth side 👍

MGibson1 commented on apps/browser/src/background/main.background.ts:
```diff
@@ -684,6 +684,7 @@ export default class MainBackground {
     await this.runtimeBackground.init();
     await this.notificationBackground.init();
     await this.commandsBackground.init();
+    (this.configService as BrowserConfigService).init();
```
We shouldn't cast here for the same reasons as in `initService`

eliykat replied:
Changed this file to use the concrete type instead of the abstraction.

MGibson1 commented on apps/browser/src/popup/services/services.module.ts:
```diff
@@ -504,6 +505,10 @@ function getBgService<T>(service: keyof MainBackground) {
         EnvironmentService,
       ],
     },
+    {
+      provide: ConfigService,
+      useExisting: ConfigServiceAbstraction,
+    },
```
This is backwards

```typescript
{
  provide: ConfigService,
  useClass: ConfigService,
  deps: [...],
},
{
  provide: ConfigServiceAbstraction,
  useExisting: ConfigService,
},
```

MGibson1 replied:
Actually since this is provided in the same way in jslib-services.module, it's likely this isn't necessary at all (thought jslib-services should be updated in this way)

eliykat replied:
I think of the abstract interface as the main DI token (with `useClass`), and registering the implementation as a bit of a limited exception which is layered on top of that (using `useExisting`). But I can see it your way as well, so I've made this change.

MGibson1 commented on libs/angular/src/services/jslib-services.module.ts:
```diff
@@ -649,6 +649,10 @@ import { AbstractThemingService } from "./theming/theming.service.abstraction";
         EnvironmentServiceAbstraction,
       ],
     },
+    {
+      provide: ConfigService,
+      useExisting: ConfigServiceAbstraction,
+    },
```
similarly backwards

cturnbull-bitwarden reviewed (dismissed):
💯

MGibson1 reviewed (dismissed):
Looks good! Sorry for all the subtleties in DI. It's complex 😓

cturnbull-bitwarden reviewed (approved):
👍
